python/lsst/alert/packet/schema.py

```diff
@@ -178,8 +178,19 @@ def validate(self, record):
         -------
         valid : `bool`
             Whether or not the data complies with the schema.
+
+        Notes
+        -----
+        Validating the against the schema requires that the fastavro cache
+        (``SCHEMA_DEFS``) be populated, but that can only be the case for one
+        version of the schema at once. Hence we populate, check for validity,
+        and flush the cache.
```

Comment:
This seems kind of gnarly, oof. Can we make a change to `fastavro` instead of needing to do this? Maybe that's just a longer-term set of changes, but fastavro really doesn't seem to be meeting our needs, based on my read of this file :)

Reply:
As the maintainer of fastavro, this definitely sounds like either a bug or some edge case that isn't being handled correctly. If someone is able to make a minimal script that reproduces this and can open an issue on fastavro then that would be appreciated. Thanks!

Reply:
Hey @scottbelden — thanks for noticing! :-)

Happy to try and put something together for you, but it might take me a few days to get organized (unless @spenczar beats me to it).

Reply:
Hi @scottbelden —

Here's an example of the problem that the above is trying to work around.

```python
from fastavro import parse_schema
from fastavro.validation import validate

# Define an schema.
schema_1 = {
    "type": "record",
    "name": "my_schema_v1",
    "fields": [
        {
            "name": "field1",
            "type": {
                "type": "record",
                "name": "my_field_type",
                "fields": [
                    {
                        "name": "sub_field",
                        "type": "string"
                    }
                ]
            }
        },
        {
            "name": "field2",
            "type": "my_field_type"
        }
    ]
}
parsed_s1 = parse_schema(schema_1)

# And some data which complies with that schema.
data_schema_1_compliant = {
    "field1": {"sub_field": "foo"}, "field2": {"sub_field": "bar"}
}
assert validate(data_schema_1_compliant, parsed_s1)  # so far so good!

# Another schema! Maybe this is an evolution of schema_2, or it's something
# completely unrelated which just happens to use the same name internally.
schema_2 = {
    "type": "record",
    "name": "my_schema_v2",
    "fields": [
        {
            "name": "field1",
            "type": {
                "type": "record",
                "name": "my_field_type",
                "fields": [
                    {
                        "name": "sub_field",
                        "type": "int"
                    }
                ]
            }
        }
    ]
}
parsed_s2 = parse_schema(schema_2)

# Neither the data nor schema_1 have changed, but this now thows a
# ValidationError.
assert validate(data_schema_1_compliant, parsed_s1)
```

As you can see, the issue is that `schema_2` defines a type with the same name as one defined in `schema_1`.

It might be possible to work around this using namespaces, but, naïvely, that seems to break down in two scenarios:

- When consuming data serialized with schemas over which we have no control (ie, schemas are managed by different entities);
- When evolving our own schemas, it's useful to be able to change types without changing names.

I would be happy to file an issue based on the above against fastavro if you'd find that helpful. However, I'm not sure if you'd regard this request as helpful; I suspect this behaviour is a design decision, rather than a bug. Please let me know if you'd like me to go ahead.

Reply:
While there definitely is some caching that is done internally, this definitely seems buggy to me. Not only what is here, but also what you alluded to in the commit about having to call `parse_schema` feels like a bug to me. While it is usually good to call `parse_schema`, the library should do that if someone forgets.

I went ahead and [created an issue in `fastavro`](https://github.com/fastavro/fastavro/issues/416) and will try to get to this over the weekend.

Thanks for help!

Reply:
Thank you, @scottbeldon! Much appreciated!